packages/shared/client/ReplayClient.ts

```diff
@@ -761,76 +804,43 @@ export class ReplayClient implements ReplayClientInterface {
-        await Promise.all(
-          locations.map(location =>
-            breakpointPositionsCache.readAsync(this, location.location.sourceId)
-          )
-        );
-      }
-
-      const analysisParams = {
-        ...rest,
-        locations,
-      };
-
-      try {
-        await analysisManager.runAnalysis(analysisParams, {
-          onAnalysisError: (error: unknown) => {
-            reject(error);
-          },
-          onAnalysisResults: analysisEntries => {
-            results.push(...analysisEntries.map(entry => entry.value));
-          },
-        });
-
-        resolve(results);
-      } catch (error) {
-        reject(error);
+  async runEvaluation(
+    opts: {
+      selector: PointSelector;
+      expression: string;
+      frameIndex?: number;
+      fullPropertyPreview?: boolean;
+      limits?: PointLimits;
+    },
+    onResults: (results: RunEvaluationResult[]) => void
+  ): Promise<void> {
+    const sessionId = this.getSessionIdThrows();
+    const runEvaluationId = String(this.nextRunEvaluationId++);
+    function onResultsWrapper(results: runEvaluationResults) {
+      if (results.runEvaluationId === runEvaluationId) {
+        onResults(results.results);
       }
-    });
-  }
-
-  streamAnalysis(
-    params: AnalysisParams,
-    handlers: {
-      onPoints?: (points: PointDescription[]) => void;
-      onResults?: (results: AnalysisEntry[]) => void;
-      onError?: (error: any) => void;
     }
-  ): { pointsFinished: Promise<void>; resultsFinished: Promise<void> } {
-    let resolvePointsFinished!: () => void;
-    const pointsFinished = new Promise<void>(resolve => (resolvePointsFinished = resolve));
-    let resolveResultsFinished!: () => void;
-    const resultsFinished = new Promise<void>(resolve => (resolveResultsFinished = resolve));
-
-    this._waitForRangeToBeLoaded(params.range || null).then(() =>
-      analysisManager.runAnalysis(params, {
-        onAnalysisPoints: handlers.onPoints,
-        onAnalysisResults: handlers.onResults,
-        onAnalysisError: error => {
-          resolvePointsFinished();
-          resolveResultsFinished();
-          handlers.onError?.(error);
-        },
-        onPointsFinished: resolvePointsFinished,
-        onFinished: resolveResultsFinished,
-      })
+
+    await this._waitForRangeToBeLoaded(
+      opts.limits?.begin && opts.limits.end
+        ? { begin: opts.limits.begin, end: opts.limits.end }
+        : null
     );
 
-    return { pointsFinished, resultsFinished };
+    addEventListener("Session.runEvaluationResults", onResultsWrapper);
+    await client.Session.runEvaluation(
+      {
+        expression: opts.expression,
+        frameIndex: opts.frameIndex,
+        fullReturnedPropertyPreview: opts.fullPropertyPreview,
+        pointLimits: opts.limits,
+        pointSelector: opts.selector,
+        runEvaluationId,
+      },
+      sessionId
+    );
+    removeEventListener("Session.runEvaluationResults", onResultsWrapper);
```

Comment:
this should really be in a `finally` so the listener doesn't leak if there's an error